File: golang/1234. 替换子串得到平衡字符串.go
func balancedString(s string) int {
    cnt:=map[byte]int{}
    for _,x:=range s{
        cnt[byte(x)]++
    }
    t:=len(s)/4
    check:=func() bool{
        if cnt['Q'] > t ||
            cnt['W'] > t ||
            cnt['E'] > t ||
            cnt['R'] > t {
            return false
        }
        return true
    }
    if check() {
        return 0
    }
    ans:=len(s)
    for l,r:=0,0;l<len(s);l++{
        for r<len(s)&&!check(){
            cnt[s[r]]--
            r++
        }
        if !check(){
            break
        }
        ans=min(ans,r-l)
        cnt[s[l]]++
    }
    return ans
}
func min(x,y int) int {
    if x>y {
        return y
    }
    return x
}

File: 刷题/cpp/1234. 替换子串得到平衡字符串.cpp
class Solution {
	private:
		int mp[26];
	public:
		int ind(char x){
			return x-'A';
		}
		bool check(int t){
			if(mp[ind('Q')]>t||mp[ind('W')]>t||mp[ind('E')]>t||mp[ind('R')]>t){
				return false;
			}
			return true;
		}
		int balancedString(string s) {
			for(auto ch:s){
				mp[ind(ch)]++;
			}
			int n=s.length();
			int t=n/4;
			if(check(t)){
				return 0;
			}
			int ans=n;
			for(int l=0,r=0;l<n;l++){
				while(r<n&&!check(t)){
					mp[ind(s[r])]--;
					r++;
				}
				if(!check(t)){
					break;
				}
				ans=min(ans,r-l);
				mp[ind(s[l])]++;
			}
			return ans;
		} 
	};